pkg/reconciler/knativeserving/ingress/kourier_test.go

```diff
@@ -19,85 +19,116 @@ package ingress
-		ns             string
-		expectedNS     string
-		expected       []v1.Container
+		instance       *servingv1alpha1.KnativeServing
+		dropLabel      bool
+		expNamespace   string
+		expServiceType string
 	}{{
-		name:           "Replaces Kourier Gateway Namespace",
-		deploymentName: DEPLOY_NAME,
-		labels: map[string]string{
-			"networking.knative.dev/ingress-provider": "kourier",
-		},
-		ns:         Kourier_GATEWAY_NS,
-		expectedNS: Expected_NS,
-		expected: []v1.Container{{
-			Name: DEPLOY_NAME,
-			Env:  []v1.EnvVar{{Name: kourierGatewayNSEnvVarKey, Value: Expected_NS}},
-		}},
+		name:           "Replaces Kourier Gateway Namespace and ServiceType",
+		instance:       servingInstance(servingNamespace, "ClusterIP"),
+		expNamespace:   servingNamespace,
+		expServiceType: "ClusterIP",
+	}, {
+		name:           "Use Kourier default service type",
+		instance:       servingInstance(servingNamespace, "" /* empty service type */),
+		expNamespace:   servingNamespace,
+		expServiceType: "LoadBalancer", // kourier GW default service type
 	}, {
-		name:           "Do Not Replace Kourier Gateway Namespace without the ingress label",
-		deploymentName: DEPLOY_NAME,
-		labels:         map[string]string{},
-		ns:             Kourier_GATEWAY_NS,
-		expectedNS:     Expected_NS,
-		expected: []v1.Container{{
-			Name: DEPLOY_NAME,
-			Env:  []v1.EnvVar{{Name: kourierGatewayNSEnvVarKey, Value: Kourier_GATEWAY_NS}},
-		}},
+		name:           "Do not transform without the ingress provier label",
+		dropLabel:      true,
+		instance:       servingInstance(servingNamespace, "ClusterIP"),
+		expNamespace:   "kourier-system", // kourier default namespace
+		expServiceType: "LoadBalancer",   // kourier GW default service type
 	}}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			u := makeUnstructuredDeployment(t, tt.deploymentName, tt.ns, tt.labels)
-			replaceKourierGWNamespace(tt.expectedNS)(u)
-			deployment := &appsv1.Deployment{}
-			err := scheme.Scheme.Convert(u, deployment, nil)
-			util.AssertEqual(t, err, nil)
-			util.AssertDeepEqual(t, deployment.Spec.Template.Spec.Containers, tt.expected)
+			client := fake.New()
+			manifest, err := mf.NewManifest("testdata/kodata/ingress/0.20/kourier.yaml", mf.UseClient(client))
+			if err != nil {
+				t.Fatalf("Failed to read manifest: %v", err)
+			}
+
+			if tt.dropLabel {
+				manifest, err = manifest.Transform(removeLabels())
+				if err != nil {
+					t.Fatalf("Failed to transform manifest: %v", err)
+				}
+			}
+
+			manifest, err = manifest.Transform(replaceGWNamespace(tt.instance))
+			if err != nil {
+				t.Fatalf("Failed to transform manifest: %v", err)
+			}
+
+			manifest, err = manifest.Transform(configureGWServiceType(tt.instance))
+			if err != nil {
+				t.Fatalf("Failed to transform manifest: %v", err)
+			}
+
+			for _, u := range manifest.Resources() {
+				if u.GetKind() == "Deployment" && u.GetName() == kourierControllerDeploymentName {
```

Comment:
Can we simplify this for loop content into one or better two functions that clearly indicate what’s going on. Especially since you have a nested for loop to deal with env. Hard to read and test IMO

Reply:
Sure, done.